panel/widgets/chat.py

```diff
@@ -0,0 +1,1656 @@
+        Extracts the contents from the entry's panel object.
+        """
+        value = entry._value_panel
+        if hasattr(value, "object"):
+            contents = value.object
+        elif hasattr(value, "objects"):
+            contents = value.objects
+        elif hasattr(value, "value"):
+            contents = value.value
+        else:
+            contents = value
+        return contents
+
+    async def _serialize_response(self, response: Any) -> ChatEntry | None:
+        """
+        Serializes the response by iterating over it and
+        updating the entry's value.
+        """
+        response_entry = None
+        if isasyncgen(response):
+            async for token in response:
+                response_entry = self._upsert_entry(token, response_entry)
+        elif isgenerator(response):
+            for token in response:
+                response_entry = self._upsert_entry(token, response_entry)
+        elif isawaitable(response):
+            response_entry = self._upsert_entry(await response, response_entry)
+        else:
+            response_entry = self._upsert_entry(response, response_entry)
+        return response_entry
+
+    async def _handle_callback(self, entry: ChatEntry) -> ChatEntry | None:
+        contents = self._extract_contents(entry)
+        response = self.callback(contents, entry.user, self)
+        response_entry = await self._serialize_response(response)
+        return response_entry
+
+    async def _schedule_placeholder(
+        self,
+        task: asyncio.Task,
+        num_entries: int,
+    ) -> None:
+        """
+        Schedules the placeholder to be added to the chat log
+        if the callback takes longer than the placeholder threshold.
+        """
+        if self.placeholder_threshold == 0:
+            return
+
+        callable_is_async = (
+            asyncio.iscoroutinefunction(self.callback) or
+            isasyncgenfunction(self.callback)
+        )
+        start = asyncio.get_event_loop().time()
+        while not task.done() and num_entries == len(self._chat_log):
+            duration = asyncio.get_event_loop().time() - start
+            if duration > self.placeholder_threshold or not callable_is_async:
+                self._chat_log.append(self._placeholder)
+                return
+            await asyncio.sleep(0.28)
+
+    async def _prepare_response(self, _) -> None:
+        """
+        Prepares the response by scheduling the placeholder and
+        executing the callback.
+        """
+        if self.callback is None:
+            return
+
+        disabled = self.disabled
+        try:
+            self.disabled = True
+            entry = self._chat_log[-1]
+            if not isinstance(entry, ChatEntry):
+                return
+
+            num_entries = len(self._chat_log)
+            task = asyncio.create_task(self._handle_callback(entry))
+            await self._schedule_placeholder(task, num_entries)
+            await task
```

Comment:
Can these be run simultaneously with `async.gather`?

Reply:
Yes I believe so; will have to confirm once async callbacks are fixed.

Reply:
It did not work, so I reverted.